### DISCUSSION TITLE
Should we add "Basic" Htmx code to core?

### DISCUSSION THREAD

**Discussion by dhgassoc** [wheels-dev/wheels#1405] [Ideas]
SURVEY: Is anyone using HTMX with cfwheels? 
How many apps (or views) have you used HTMX with? on internal intranet or on web?

I think we should add "basic" functionality for HTMX in the core code.   Like chwheels does for isAjax(), add basic  isHtmx() support.  

To get HTMX + CSRF to work on Adobe-2016 in my browser, I had to modify $setAuthenticityToken().  I don't know if or how isAjax() is used elsewhere in core, and if other core code needs ` (isAjax() || isHtmx() ) ` added. 

Should we add needed htmx+CSRF explicit support for isHtmx() in csrf.cfm's  $setAuthenticityToken() in /wheels/controller/csrf.cfm, to;
``` 
public function $setAuthenticityToken() {	
	if (!$isVerifiedRequest() && (isAjax() ||      isHtmx()      ) ) {	
		if (StructKeyExists(request.$wheelsHeaders, "X-CSRF-Token")) {	
			params.authenticityToken = request.$wheelsHeaders["X-CSRF-Token"];	
		}	
	}	
}
 ```

Does it make sense to add a "htmx" format response type, line cfwheels does for xml, json, html?  add support for HTMX in provides() ?

Do you think that "basic" HTMX support should be added to core?

**Comment by bpamiri** [comment]
My thought on this is that we should first identify all the areas that could use better HTMX support. Then see if they could be addressed with a plugin or if it is so vast that adding to the core makes more sense. 

I'm not sure that a HTMX attribute in provides() makes sense cause at the end of the day HTMX just requires HTML to be returned which is what a Wheels view does inherently. Areas that I've often struggled with and have come up with various bandaids in my code are when I try to bypass the view by just doing renderText() in the controller. This is okay for simple stuff but if you try to get too fancy with an OOB directive imbedded in the returned text then things get a little hairy. Or another place might be how flash messages might work if you are returning an HTMX response. Or more HTMX integration in the Wheels CLI for debugging purposes. 

I guess what I'm saying is let's try to collect a list of all the pain points people are having with HTMX and how better integration might resolve them. We can look at Django for pointers in what they built and see if it makes sense for us to implement.

**Comment by chapmandu** [comment]
My opinion is a firm no. This type of integration belongs in a plugin. 

I LOVE HTMX, but wheels should not be tied to any JavaScript library. 

Where does it stop? alpine.js, hyperscript, tailwind, turbo, stimulus..

**Comment by mikegrogan** [comment]
Haven't used htmx yet, but am interested in learning more about how it could integrate with CFWheels. Along with alpine.js, etc. Agreed with the plugin approach though. Need to keep improving the core first.

**Comment by bpamiri** [comment]
I agree on the plugin approach. My point was from the perspective of the arc of time. First these items are added as a plugin, over time, if it become ubiquitous and the entire community is using a particular plugin, then we look to integrate it into the core. 

It is fairly simple to extend or override an internal function via the plugin functionality so this approach shouldn't cause any roadblocks.
